### PR TITLE
fix: Include hdfs principal names in discovery ConfigMap

### LINKED ISSUE #419: Discovery configmap incomplete when Kerberos is enabled
### Affected version

23.7.0

### Current and expected behavior

When Kerberos is enabled on HDFS a number of properties must be added to core-site.xml to enable clients to determine the correct kerberos principal for the services. Without this the client will be unable to determine which principal to use when requesting a Kerberos ticket and you see the following error in the client logs:

```
java.lang.IllegalArgumentException: Failed to specify server's Kerberos principal name, while invoking ClientNamenodeProtocolTranslatorPB.getListing over poc-hdfs-namenode-default-0.poc-hdfs-namenode-default.default.svc.cluster.local/10.244.3.19:8020
```

### Possible solution

The following properties should be added to core-site.xml in the  HDFS discovery configMap when Kerberos is enabled:

- dfs.datanode.kerberos.principal
- dfs.namenode.kerberos.principal
- dfs.web.authentication.kerberos.principal

### Additional context

_No response_

### Environment

_No response_

### Would you like to work on fixing this bug?

None

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Fixes https://github.com/stackabletech/hdfs-operator/issues/419
Fixed while working on https://github.com/stackabletech/hbase-operator/issues/285.
Strangely HBase works without the principals being configured...

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: docs/modules/hdfs/pages/discovery.adoc [hunk 1]
```diff
@@ -36,4 +36,13 @@ The `ConfigMap` data values are formatted as Hadoop XML files which allows simpl
...
 
 `hdfs-site.xml`::
-Contains the `dfs.namenode.*` properties for `rpc` and `http` addresses for the `namenodes` as well as the `dfs.nameservices` property which defaults to `hdfs://{clusterName}/`.
\ No newline at end of file
+Contains the `dfs.namenode.*` properties for `rpc` and `http` addresses for the `namenodes` as well as the `dfs.nameservices` property which defaults to `hdfs://{clusterName}/`.
+
+=== Kerberos
+In case Kerberos is enabled according the xref:usage-guide/security.adoc[security documentation] the discovery will also include the information that clients must authenticate themselves using Kerberos.
+
+Some Kerberos-related configs require the environmental variable `KERBEROS_REALM` to be set (e.g. using `export KERBEROS_REALM=$(grep -oP 'default_realm = \K.*' /stackable/kerberos/krb5.conf)`).
```
Comment (fhennig) [inline docs/modules/hdfs/pages/discovery.adoc]:
```suggestion
Some Kerberos-related configuration settings require the environment variable `KERBEROS_REALM` to be set (e.g. using `export KERBEROS_REALM=$(grep -oP 'default_realm = \K.*' /stackable/kerberos/krb5.conf)`).
```

File: docs/modules/hdfs/pages/discovery.adoc [hunk 2]
```diff
@@ -36,4 +36,13 @@ The `ConfigMap` data values are formatted as Hadoop XML files which allows simpl
...
+When you want to use the discovery configmap outside of Stackable services, you need to provide this environment variable!
```
Comment (fhennig) [inline docs/modules/hdfs/pages/discovery.adoc]:
```suggestion
When you want to use the discovery ConfigMap outside of Stackable services, you need to provide this environment variable!
```

**Review by fhennig — CHANGES_REQUESTED** [review]
code looks good, but can we get a test maybe?

File: rust/operator/src/kerberos.rs [hunk 3]
```diff
@@ -59,22 +59,10 @@ impl CoreSiteConfigBuilder {
...
-            // Only Namenode and another JournalNode may access this servlet
-            //
-            // After we have switched to using the following principals everything worked without problems
+            let principal_host_part = principal_host_part(hdfs_name, hdfs_namespace);
 
-            let principal_host_part =
-                format!("{hdfs_name}.{hdfs_namespace}.svc.cluster.local@${{env.KERBEROS_REALM}}");
             self.add("hadoop.security.authentication", "kerberos")
-                .add("hadoop.registry.kerberos.realm", "${env.KERBEROS_REALM}")
+                // Not adding hadoop.registry.kerberos.realm, as it seems to not be used by our customers
```
Comment (maltesander) [inline rust/operator/src/kerberos.rs]: Does it hurt though :) ?

File: rust/operator/src/container.rs [hunk 4]
```diff
@@ -317,9 +329,13 @@ impl ContainerConfig {
     /// - Namenode ZooKeeper fail over controller (ZKFC)
     /// - Datanode main process
     /// - Journalnode main process
+
+    #[allow(clippy::too_many_arguments)]
     fn main_container(
         &self,
         hdfs: &HdfsCluster,
+        hdfs_name: &str,
+        hdfs_namespace: &str,
```
Comment (maltesander) [inline rust/operator/src/container.rs]: Do we really want to add extra parameters for something i can get from the `HdfsCluster`? 
Same for other methods.
Id prefer to just impl this on the cluster object and call it?

Comment (sbernauer) [reply rust/operator/src/container.rs]: Created https://github.com/stackabletech/hdfs-operator/pull/451 based on this review comment, please have a look there

**Comment by sbernauer** [comment]
Superseded by https://github.com/stackabletech/hdfs-operator/pull/451

**Pull request closed without merging**